# RAG 기반 Customer Service AI 에이전트 개발 - 과제

---

## 학습 개요

### 학습 주제
- **청킹 전략 비교 및 선택**: chunk_size, overlap 설정에 따른 검색 결과 품질 차이를 실험적으로 비교
- **Retriever 설정 최적화**: top_k 등 검색 파라미터 조정에 따른 RAG 응답 품질 변화 분석
- **RAG 파이프라인 응용**: 실습에서 구현한 파이프라인을 기반으로 설정을 최적화하고 근거를 제시

### 학습 목표
- chunk_size를 변경했을 때 검색 결과가 어떻게 달라지는지 비교하고 차이점을 설명할 수 있다
- top_k 값을 변경했을 때 RAG 응답 품질이 어떻게 달라지는지 비교하고 차이점을 설명할 수 있다
- Yes24 고객 서비스 시나리오에 적합한 RAG 설정(chunk_size, overlap, top_k)을 선택하고 논리적 근거를 제시할 수 있다

### 핵심 개념
- **청킹 전략(Chunking Strategy)**: chunk_size(청크 크기)와 overlap(겹침)이 검색 결과에 미치는 영향. 작은 청크는 세밀한 검색이 가능하나 문맥 부족, 큰 청크는 문맥 풍부하나 노이즈 포함 위험
- **Retriever 파라미터**: top_k(반환 문서 수) 등 검색 설정이 RAG 응답 품질에 미치는 영향. 적은 k는 정보 누락 위험, 많은 k는 관련 없는 정보 포함 위험
- **설정 비교/분석**: 실험 결과를 바탕으로 상황에 맞는 최적 설정을 판단하고 근거를 제시하는 역량

### 선행 지식
- **4-1 실습 완료 필수**: RAG 파이프라인(임베딩, Vector Store, Retriever, LangGraph) 구현 경험
- **Python 프로그래밍**: 함수 호출, 딕셔너리/리스트 조작, 출력 결과 해석 (기초~중급)

### 실습-과제 연결
- 이 과제는 **4-1 실습에서 구현한 RAG 파이프라인을 기반**으로 합니다.
- 실습이 **구현 중심**이었다면, 과제는 **비교/분석 중심**으로 설정 선택 역량을 평가합니다.
- **약한 변형 (조건 변화)**: 실습에서 사용한 chunk_size=300 대신 200, 500으로 변경하여 비교
- **마무리 변형 (해석/분석)**: 실험 결과를 바탕으로 최적 설정 선택 및 근거 서술

### Further Reading
- [Retrieval-Augmented Generation for Knowledge-Intensive NLP Tasks](https://arxiv.org/abs/2005.11401): RAG 원본 논문 (Lewis et al., 2020). RAG의 이론적 배경과 효과 이해
- [Dense Passage Retrieval for Open-Domain QA](https://arxiv.org/abs/2004.04906): 의미 기반 검색의 핵심 기술인 Dense Retrieval 논문
- [Chunking Strategies for LLM Applications](https://www.pinecone.io/learn/chunking-strategies/): 다양한 청킹 전략(Fixed-size, Semantic, Recursive)의 장단점 비교
- [LangChain Retriever 공식 문서](https://python.langchain.com/docs/modules/data_connection/retrievers/): Retriever 설정 옵션 및 고급 기능

### Open-ended Mission
- Semantic Chunking 적용: RecursiveCharacterTextSplitter 대신 의미 단위로 청킹하는 방식 적용해보기
- Reranker 추가: 검색 결과를 Reranker로 재정렬하여 정확도 개선해보기
- 다른 문서 적용: Yes24 문서 대신 다른 도메인 문서에 동일 파이프라인 적용해보기

---

## 과제 구성

### 학습 방향

- **실습 구성 방식**
  - 실습에서 구현한 파이프라인 기반, 구성 요소 변경 후 비교/분석합니다. 각 TODO에서 선택 이유와 근거를 함께 작성합니다.


- **실행 환경**
  - Python 3.10 이상

- **Required Package**
  - `langchain>=0.3.0`
  - `langchain-openai>=0.2.0`
  - `langchain-community>=0.3.0`
  - `langgraph>=0.2.0`
  - `chromadb>=0.4.0`
  - `tiktoken>=0.5.0`
  - `python-dotenv>=1.0.0`
  - `pymupdf>=1.24.0`

- **Step 요약**
  - **Step 1 (5분)**: 환경 설정 + 실습 복습 — 실습 완료 전제, 환경 확인
  - **Step 2 (15분)**: 청킹 전략 변경 및 비교 — chunk_size, overlap 변경 → 결과 비교
  - **Step 3 (15분)**: Retriever 설정 변경 및 비교 — top_k 등 변경 → 검색 품질 비교
  - **Step 4 (10분)**: 결과 분석 및 근거 서술 — 설정 선택 이유 정리

### 문제 설명

- **문제 개요**: 이 과제는 **RAG 파이프라인의 구성 요소를 판단하고 선택하는 역량**을 평가합니다. 학습자는 **청킹 전략, Retriever 설정 등을 변경**하며 결과를 비교하고, 최종적으로 **선택한 구성에 대한 논리적 근거를 제시**할 수 있어야 합니다.
- **요구사항 요약**
  - **청킹 전략 비교**: chunk_size 변경에 따른 검색 결과 차이 분석
  - **Retriever 설정 비교**: top_k 변경에 따른 RAG 응답 품질 비교
  - **근거 기반 의사결정**: 최적 설정 선택 및 논리적 근거 서술

### 데이터셋 개요 및 저작권 정보

- **데이터셋 명**: Yes24 서비스 혜택 문서
- **데이터셋 개요**: 실습과 동일한 Yes24 서비스 문서를 사용하여 RAG 파이프라인 설정 비교 실험을 수행합니다.
- **사용 목적**: RAG 파이프라인 설정 비교 실험
- **저작권/출처**: Yes24 공개 서비스 안내 (교육 목적 사용)
- **주의사항**: 실습 코드를 기반으로 진행하므로, 실습 완료 후 과제를 수행해야 합니다.

### 학습 문제

- **Step 2 — 청킹 전략 변경 및 비교**
  - **TODO 1**: chunk_size 값 변경하여 비교 *(연결: Chunking, chunk_size 트레이드오프)*
  - **TODO 2**: 검색 결과 비교 및 차이점 분석 *(연결: 청킹 전략 선택 기준)*
  - **1줄 요약**: chunk_size에 따른 검색 결과 차이를 관찰하고 분석한다.
- **Step 3 — Retriever 설정 변경 및 비교**
  - **TODO 3**: top_k 값 변경하여 비교 *(연결: Retriever, top_k 설정)*
  - **TODO 4**: 검색 결과 비교 및 차이점 분석 *(연결: 검색 품질과 top_k의 관계)*
  - **1줄 요약**: top_k에 따른 RAG 응답 품질 차이를 관찰하고 분석한다.
- **Step 4 — 결과 분석 및 근거 서술**
  - **TODO 5**: 최종 설정 선택 및 근거 서술 *(연결: 의사결정 역량, 트레이드오프 분석)*
  - **1줄 요약**: 최적의 RAG 설정을 선택하고 논리적 근거를 제시한다.


---
# Step 1: 환경 설정 + 실습 복습

### Setup: 환경 설정 및 기본 파이프라인 구성

실습에서 배운 RAG 파이프라인을 빠르게 구성합니다.
이 코드는 **제공되며**, 학습자는 실행만 하면 됩니다.

In [2]:
# 패키지 설치
# %%capture
%pip install "langchain>=0.3.0" "langchain-openai>=0.2.0" "langchain-community>=0.3.0" "langgraph>=0.2.0" "langchain-text-splitters>=0.3.0" "chromadb>=0.5.0" "tiktoken>=0.7.0" "pymupdf>=1.24.0" "python-dotenv>=1.0.0" -q

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import os
import glob
import random
import torch
import warnings
from dotenv import load_dotenv

warnings.filterwarnings("ignore")

# .env 파일에서 환경변수 로드
load_dotenv()

# 재현성을 위한 시드 설정
random.seed(42)
torch.manual_seed(42)

# API 키 확인 (.env 미설정 시 직접 입력)
if not os.environ.get("GMS_KEY"):
    os.environ["GMS_KEY"] = input("🔑 GMS_KEY를 입력하세요: ")
    
print("✅ 환경 설정 완료")

✅ 환경 설정 완료


In [4]:
from langchain_openai import ChatOpenAI, OpenAIEmbeddings

llm = ChatOpenAI(
    model="gpt-5-mini",
    api_key=os.environ["GMS_KEY"],
    base_url="https://gms.ssafy.io/gmsapi/api.openai.com/v1/"
)

embeddings = OpenAIEmbeddings(
    model="text-embedding-3-small",
    api_key=os.environ["GMS_KEY"],
    base_url="https://gms.ssafy.io/gmsapi/api.openai.com/v1/"
)

In [5]:
# 모델 연결 테스트
response = llm.invoke("Yes24에서 배송 지연 시 보상 절차는 어떻게 되나요?")
print("LLM 응답:")
print(response.content[:200] + "...")
print(f"\n✅ 모델 연결 확인 완료 ({llm.model_name})")

LLM 응답:
간단히 정리하면, Yes24에서 배송이 지연되면 다음 절차로 보상(환불·배송비 환급·쿠폰 등)을 요청할 수 있습니다.

1) 우선 상태 확인
- My YES24(또는 주문/배송조회)에서 주문번호·배송조회(운송장) 상태를 확인하세요.
- 안내된 배송예정일을 넘어섰는지, 운송장 추적상 배송지연(집하·배송중 오류 등)인지 확인합니다.

2) 판매자/배송사 확인
...

✅ 모델 연결 확인 완료 (gpt-5-mini)


In [6]:
from langchain_community.document_loaders import PyMuPDFLoader
# 데이터 로드 (실습과 동일)
DATA_DIR = "data/"

# 경로 안내:
# 이 노트북과 같은 디렉토리에 "data/" 폴더를 생성하고
# Yes24 PDF 문서들을 다음과 같은 구조로 배치해주세요:
#   - data/
#     - yes24_서비스혜택.pdf
#     - yes24_배송정책.pdf
#     - (기타 관련 PDF 파일들)

pdf_files = glob.glob(DATA_DIR + "*.pdf")

if not pdf_files:
    print(f"⚠️ 경고: {DATA_DIR} 디렉토리에서 PDF 파일을 찾을 수 없습니다.")
    print("데이터 디렉토리 구조를 확인해주세요.")

all_documents = []
for pdf_path in pdf_files:
    loader = PyMuPDFLoader(pdf_path)
    docs = loader.load()
    for doc in docs:
        doc.metadata["source_file"] = os.path.basename(pdf_path)
    all_documents.extend(docs)

print(f"📚 총 {len(all_documents)}개 페이지 로드 완료")
print(f"📁 문서 목록: {set([doc.metadata.get('source_file') for doc in all_documents])}")

📚 총 8개 페이지 로드 완료
📁 문서 목록: {'배송지연 보상제도_서비스 혜택 - 예스24.pdf', '신규 회원 혜택_서비스 혜택 - 예스24.pdf', '영원한 YES포인트_서비스 혜택 - 예스24.pdf', '총알배송_서비스 혜택 - 예스24.pdf', '무료배송 추가적립_서비스 혜택 - 예스24.pdf', '매장 픽업 서비스_서비스 혜택 - 예스24.pdf', '도서 품절 보상제도 - 예스24.pdf', '제휴 할인카드_서비스 혜택 - 예스24.pdf'}


### Concept Check: 실습 복습

실습에서 배운 RAG 파이프라인의 핵심 구성요소:

1. **청킹(Chunking)**: 문서를 적절한 크기로 분할
   - `chunk_size`: 청크의 최대 크기
   - `chunk_overlap`: 청크 간 겹침 (문맥 유지)

2. **Vector Store**: 임베딩 벡터 저장 및 검색
   - ChromaDB 사용

3. **Retriever**: 유사한 문서 검색
   - `k`: 반환할 문서 수

4. **RAG Chain**: 검색 + 생성 통합

이 과제에서는 **청킹 설정**과 **Retriever 설정**을 변경하며 **결과를 비교**합니다.

---
# Step 2: 청킹 전략 변경 및 비교

### Concept Check: chunk_size와 overlap의 영향

| 설정 | 작은 값 | 큰 값 |
|------|---------|-------|
| chunk_size | 세밀한 검색, 문맥 부족 | 넓은 문맥, 검색 정확도 하락 |
| overlap | 문맥 단절 위험 | 중복 정보, 비용 증가 |

**실험 목표**: 다양한 설정으로 검색 품질이 어떻게 달라지는지 비교

### Guided Build: 청킹 및 Vector Store 생성 함수

아래 함수를 사용하여 다양한 설정으로 실험합니다.

In [7]:
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_community.vectorstores import Chroma
import tiktoken

def create_vectorstore_with_chunking(documents, chunk_size, chunk_overlap, collection_name):
    """
    주어진 청킹 설정으로 Vector Store를 생성합니다.
    
    Args:
        documents: 원본 문서 리스트
        chunk_size: 청크 크기
        chunk_overlap: 청크 간 겹침
        collection_name: ChromaDB 컬렉션 이름
    
    Returns:
        vectorstore, chunked_documents
    """
    # 청킹
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=chunk_size,
        chunk_overlap=chunk_overlap,
        length_function=len,
        separators=["\n\n", "\n", ".", " ", ""]
    )
    chunked_docs = text_splitter.split_documents(documents)
    
    # 임베딩 모델의 토큰 제한 안전장치 (text-embedding-3-small: 최대 8,191 토큰)
    MAX_TOKENS = 8000
    enc = tiktoken.encoding_for_model("text-embedding-3-small")
    for doc in chunked_docs:
        tokens = enc.encode(doc.page_content)
        if len(tokens) > MAX_TOKENS:
            doc.page_content = enc.decode(tokens[:MAX_TOKENS])
    
    # Vector Store 생성
    vectorstore = Chroma.from_documents(
        documents=chunked_docs,
        embedding=embeddings,
        collection_name=collection_name
    )
    
    return vectorstore, chunked_docs

print("✅ 함수 정의 완료")

✅ 함수 정의 완료


### TODO 1: chunk_size 값 변경하여 비교

- **요구사항**: chunk_size를 200과 500으로 각각 설정하여 검색 결과 비교
- **입력**: chunk_size 값 (200, 500)
- **출력**: 각 설정별 검색 결과
- **힌트**: `create_vectorstore_with_chunking()` 함수 사용

In [8]:
# 1. chunk_size 값 변경하여 비교

# ========== 여기에 코드를 작성하세요 ==========
# 설정 A: 작은 chunk_size (200)
# 힌트: create_vectorstore_with_chunking(all_documents, chunk_size=200, chunk_overlap=20, collection_name="...")

vectorstore_small, chunks_small = create_vectorstore_with_chunking(all_documents, chunk_size=200, chunk_overlap=20, collection_name="small_chunk")  # TODO: 이 줄을 수정하세요

# 설정 B: 큰 chunk_size (500)
vectorstore_large, chunks_large = create_vectorstore_with_chunking(all_documents, chunk_size=500, chunk_overlap=50, collection_name="large_chunk")  # TODO: 이 줄을 수정하세요

# ============================================

if chunks_small and chunks_large:
    print(f"📊 청킹 결과 비교:")
    print(f"  - chunk_size=200: {len(chunks_small)}개 청크")
    print(f"  - chunk_size=500: {len(chunks_large)}개 청크")
else:
    print("⚠️ 코드를 완성해주세요!")

📊 청킹 결과 비교:
  - chunk_size=200: 79개 청크
  - chunk_size=500: 33개 청크


In [9]:
# 검색 비교
if vectorstore_small and vectorstore_large:
    retriever_small = vectorstore_small.as_retriever(search_kwargs={"k": 3})
    retriever_large = vectorstore_large.as_retriever(search_kwargs={"k": 3})

    test_query = "배송이 늦으면 어떻게 보상받나요?"

    print(f"🔍 질문: {test_query}\n")

    # 작은 chunk로 검색
    results_small = retriever_small.invoke(test_query)
    print("📄 [chunk_size=200 검색 결과]")
    for i, doc in enumerate(results_small, 1):
        print(f"  결과 {i} (길이: {len(doc.page_content)}자): {doc.page_content[:100]}...")

    print()

    # 큰 chunk로 검색
    results_large = retriever_large.invoke(test_query)
    print("📄 [chunk_size=500 검색 결과]")
    for i, doc in enumerate(results_large, 1):
        print(f"  결과 {i} (길이: {len(doc.page_content)}자): {doc.page_content[:100]}...")

🔍 질문: 배송이 늦으면 어떻게 보상받나요?

📄 [chunk_size=200 검색 결과]
  결과 1 (길이: 196자): 설날, 추석 등 명절 성수기로 택배 물량이 급증하여 배송이 지연되는 경우 제외
보상 서비스가 불가능할 시, 별도의 공지를 한 경우 제외 (예: 재고조사, 시스템 점검, 센터 이전 ...
  결과 2 (길이: 173자): 싸다
빠르다
믿을 수 있다
예스24에서 주문하신 아침/당일/하루 배송 상품이 주문 완료 시점에 안내된 도착 예상 일보다 지연된 경우 주문 건당 YES포인트 2,000원을 보상해 드...
  결과 3 (길이: 193자): 배송지역변경 버튼 설정 후 도착예정일 확인 부탁드립니다
분철 옵션 선택 시 당일/하루배송이 불가능합니다.
지역별 하루배송 확률은 최근 30일 배송 데이터에 근거한 정보이므로, 상황...

📄 [chunk_size=500 검색 결과]
  결과 1 (길이: 499자): 싸다
빠르다
믿을 수 있다
예스24에서 주문하신 아침/당일/하루 배송 상품이 주문 완료 시점에 안내된 도착 예상 일보다 지연된 경우 주문 건당 YES포인트 2,000원을 보상해 드...
  결과 2 (길이: 491자): 싸다
빠르다
믿을 수 있다
예스24에서 배송하는 도서(잡지, 중고도서, 직수입 도서 제외)가 일시품절, 품절, 절판, 판매금지, 미출간 상태로 변경된 경우, 주문취소 후 품절 상품...
  결과 3 (길이: 498자): 당일배송
자정
하루배송
0시
12시
당일배송
자정
하루배송
0시
13시
당일배송
자정
하루배송
0시
13시
당일배송
자정
하루배송
0시
12시
당일배송
자정
하루배송
0시
12시
...


### TODO 2(서술형): 검색 결과 비교 및 차이점 분석

- **요구사항**: 위 실험 결과를 바탕으로 chunk_size에 따른 차이점 분석
- **작성**: 아래 마크다운 셀에 분석 내용 작성 (2~3줄)

#### TODO 2 답안: chunk_size 차이점 분석

**관찰 결과:**
- chunk_size=200: (  결과 1 (길이: 196자): 설날, 추석 등 명절 성수기로 택배 물량이 급증하여 배송이 지연되는 경우 제외
보상 서비스가 불가능할 시, 별도의 공지를 한 경우 제외 (예: 재고조사, 시스템 점검, 센터 이전 ...
  결과 2 (길이: 173자): 싸다
빠르다
믿을 수 있다
예스24에서 주문하신 아침/당일/하루 배송 상품이 주문 완료 시점에 안내된 도착 예상 일보다 지연된 경우 주문 건당 YES포인트 2,000원을 보상해 드...
  결과 3 (길이: 193자): 배송지역변경 버튼 설정 후 도착예정일 확인 부탁드립니다
분철 옵션 선택 시 당일/하루배송이 불가능합니다.
지역별 하루배송 확률은 최근 30일 배송 데이터에 근거한 정보이므로, 상황...)
- chunk_size=500: (결과 1 (길이: 499자): 싸다
빠르다
믿을 수 있다
예스24에서 주문하신 아침/당일/하루 배송 상품이 주문 완료 시점에 안내된 도착 예상 일보다 지연된 경우 주문 건당 YES포인트 2,000원을 보상해 드...
  결과 2 (길이: 491자): 싸다
빠르다
믿을 수 있다
예스24에서 배송하는 도서(잡지, 중고도서, 직수입 도서 제외)가 일시품절, 품절, 절판, 판매금지, 미출간 상태로 변경된 경우, 주문취소 후 품절 상품...
  결과 3 (길이: 498자): 당일배송
자정
하루배송
0시
12시
당일배송
자정
하루배송
0시
13시
당일배송
자정
하루배송
0시
13시
당일배송
자정
하루배송
0시
12시
당일배송
자정
하루배송
0시
12시
...)

**결론**: `chunk_size`가 작은 경우(200) 질의와 직결된 핵심 정보만을 정밀하게 추출할 수 있으나 문맥이 단절될 위험이 있고, `chunk_size`가 큰 경우(500) 문서의 전반적인 문맥과 풍부한 정보를 담을 수 있지만 불필요한 노이즈 데이터가 함께 포함될 수 있습니다. 따라서 RAG 구현 시 질문의 복잡도와 문서의 특성에 맞춰 적절한 청크 크기를 산정해야 합니다.

In [ ]:
# (서술형 TODO) 위 마크다운 셀에 답안을 작성하세요.
# 이 셀은 실행하지 않아도 됩니다.

---
# Step 3: Retriever 설정 변경 및 비교

### Concept Check: top_k의 영향

**top_k**: 검색 시 반환할 문서 개수

| top_k 값 | 특징 |
|----------|------|
| 작음 (1~2) | 가장 관련성 높은 문서만, 정보 누락 위험 |
| 중간 (3~5) | 균형 잡힌 검색 |
| 큼 (5+) | 다양한 정보, 관련 없는 내용 포함 위험 |

### Guided Build: RAG 체인 생성 함수

아래 함수를 사용하여 다양한 top_k로 RAG 결과를 비교합니다.

In [11]:
from langchain_core.prompts import ChatPromptTemplate
# 기본 설정으로 Vector Store 생성 (비교 실험용)
vectorstore_base, _ = create_vectorstore_with_chunking(
    documents=all_documents,
    chunk_size=300,
    chunk_overlap=50,
    collection_name="yes24_base"
)

# RAG 프롬프트
rag_prompt = ChatPromptTemplate.from_messages([
    ("system", """당신은 온라인 서점 'Yes24'의 고객 서비스 상담원입니다.
다음 검색된 자료를 참고하여 고객의 질문에 정확하고 친절하게 답변해주세요.

[검색된 자료]
{context}

답변 규칙:
1. 검색된 자료를 기반으로 답변하세요
2. 자료에 없는 내용은 추측하지 마세요"""),
    ("human", "{question}")
])

def format_docs(docs):
    return "\n\n".join([doc.page_content for doc in docs])

def run_rag(retriever, question):
    """
    주어진 retriever와 질문으로 RAG를 수행합니다.
    """
    # 관련 문서 검색
    docs = retriever.invoke(question)
    context = format_docs(docs)
    
    # 프롬프트 생성 후 LLM 직접 호출
    messages = rag_prompt.format_messages(context=context, question=question)
    response = llm.invoke(messages)
    
    return response.content

def create_rag_components(vectorstore, k):
    """
    주어진 top_k 설정으로 RAG 컴포넌트를 생성합니다.
    """
    retriever = vectorstore.as_retriever(search_kwargs={"k": k})
    return retriever

print("✅ 함수 정의 완료")

✅ 함수 정의 완료


### TODO 3: top_k 값 변경하여 비교

- **요구사항**: top_k를 2와 5로 각각 설정하여 RAG 응답 비교
- **입력**: top_k 값 (2, 5)
- **출력**: 각 설정별 RAG 응답
- **힌트**: `create_rag_components()` 함수 사용

In [12]:
# TODO: top_k 값 변경하여 비교

# ========== 여기에 코드를 작성하세요 ==========
# 설정 A: 작은 top_k (2)
# 힌트: create_rag_components(vectorstore_base, k=2)

retriever_k2 = create_rag_components(vectorstore_base, k=2)  # 1. 이 줄을 수정하세요

# 설정 B: 큰 top_k (5)
retriever_k5 = create_rag_components(vectorstore_base, k=5)  # 2. 이 줄을 수정하세요

# ============================================

if retriever_k2 and retriever_k5:
    print("✅ Retriever 생성 완료 (k=2, k=5)")
else:
    print("⚠️ 코드를 완성해주세요!")

✅ Retriever 생성 완료 (k=2, k=5)


In [13]:
# 검색 결과 비교
if retriever_k2 and retriever_k5:
    test_query = "YES포인트 적립과 사용 방법을 알려주세요"

    print(f"🔍 질문: {test_query}\n")
    print("="*60)

    # k=2 검색 결과
    results_k2 = retriever_k2.invoke(test_query)
    print(f"📄 [top_k=2] 검색된 문서 수: {len(results_k2)}개")
    for i, doc in enumerate(results_k2, 1):
        print(f"  문서 {i}: {doc.page_content[:80]}...")

    print()

    # k=5 검색 결과
    results_k5 = retriever_k5.invoke(test_query)
    print(f"📄 [top_k=5] 검색된 문서 수: {len(results_k5)}개")
    for i, doc in enumerate(results_k5, 1):
        print(f"  문서 {i}: {doc.page_content[:80]}...")

🔍 질문: YES포인트 적립과 사용 방법을 알려주세요

📄 [top_k=2] 검색된 문서 수: 2개
  문서 1: 싸다
빠르다
믿을 수 있다
YES포인트는 유효기간이 없습니다.
YES포인트는 영원합니다.
YES포인트는 상품 구매나 커뮤니티 활동에 대한 보상으...
  문서 2: KB국민카드 결제 시 포인트리 0.2% 적립 (무이자 할부, 선불/기프트카드, 일부 제휴카드, 포인트 사용금액은 제외)
문의 : KB국민카드 고...

📄 [top_k=5] 검색된 문서 수: 5개
  문서 1: 싸다
빠르다
믿을 수 있다
YES포인트는 유효기간이 없습니다.
YES포인트는 영원합니다.
YES포인트는 상품 구매나 커뮤니티 활동에 대한 보상으...
  문서 2: KB국민카드 결제 시 포인트리 0.2% 적립 (무이자 할부, 선불/기프트카드, 일부 제휴카드, 포인트 사용금액은 제외)
문의 : KB국민카드 고...
  문서 3: 싸다
빠르다
믿을 수 있다
최저가격 보장은 기본!
예스24 카드로 추가 적립 혜택도 누리세요!
서비스별로 적립 및 할인 정책이 변경될 수 있으니...
  문서 4: 자세히보기
제휴카드 포인트 할인/적립
신한카드 마이신한포인트
도서, CD/LP, DVD/Blu-ray, 문구/GIFT 결제 시 1점부터 결제 가...
  문서 5: 보상금액
품절 취소 상품에 한해 상품별로 판매가에 따라 차등 지급
판매가 5,000원 이상은 YES포인트 1,000원 지급
판매가 5,000원 ...


In [14]:
# RAG 응답 비교
if retriever_k2 and retriever_k5:
    test_query = "YES포인트 적립과 사용 방법을 알려주세요"
    
    print(f"🔍 질문: {test_query}")
    print("\n" + "="*60)

    # k=2 응답
    print("\n🤖 [top_k=2] RAG 응답:")
    response_k2 = run_rag(retriever_k2, test_query)
    print(response_k2)

    print("\n" + "="*60)

    # k=5 응답
    print("\n🤖 [top_k=5] RAG 응답:")
    response_k5 = run_rag(retriever_k5, test_query)
    print(response_k5)

🔍 질문: YES포인트 적립과 사용 방법을 알려주세요


🤖 [top_k=2] RAG 응답:
안녕하세요. YES24 고객센터입니다. YES포인트 적립과 사용에 대해 안내드리겠습니다.

적립
- YES포인트는 상품 구매나 커뮤니티 활동에 대한 보상으로 회원님께 쌓아드리는 포인트입니다.
- 상품별로 적립률이 정해져 있으며, 가격 할인 쿠폰과 별도로 YES포인트가 지급됩니다.
- 도서, eBook, CD/LP, DVD/Blu-ray, 문구/GIFT 카테고리에서 5만원 이상 주문하시면 추가 적립 혜택이 제공됩니다.
- YES포인트는 유효기간이 없습니다. (YES포인트는 영원합니다.)

사용
- YES포인트가 5,000원 이상 모이면 YES머니로 전환하여 현금과 동일하게 사용하실 수 있습니다.

카드 관련 적립/사용 안내 (참고)
- KB국민카드 결제 시 포인트리 0.2% 적립(무이자 할부, 선불/기프트카드, 일부 제휴카드, 포인트 사용금액은 제외) — 문의: KB국민카드 고객센터 1588-1688
- BC카드 탑포인트: 도서·CD/LP·DVD/Blu-ray·문구/GIFT 쇼핑 시 1점부터 결제 가능(결제 시 BC카드 선택 후 BC 탑포인트 사용 동의). BC카드 결제 시 탑포인트 1% 적립(선불/기프트카드, 일부 제휴카드, 포인트 사용금액은 제외) — 문의: BC카드 고객센터 1588-4000
- (검색 자료에 따르면 씨티카드 관련으로는 ‘씨티포인트’가 있습니다.)

추가로 궁금하신 점(예: 보유 포인트 확인 방법, YES머니 전환 절차 등)이 있으시면 말씀해 주세요. 검색된 자료 범위 내에서 자세히 안내해 드리겠습니다.


🤖 [top_k=5] RAG 응답:
안녕하세요 고객님, 예스24 고객서비스입니다. YES포인트 적립·사용 관련 요약 안내드립니다.

1) YES포인트란?
- 상품 구매나 커뮤니티 활동에 대한 보상으로 적립해 드리는 포인트입니다.
- 유효기간이 없습니다. (YES포인트는 영원합니다.)

2) 적립 방법
- 상품 구매 시 상품별로 적립됩니다. 상품마

### TODO 4(서술형): 검색 결과 비교 및 차이점 분석

- **요구사항**: 위 실험 결과를 바탕으로 top_k에 따른 차이점 분석
- **작성**: 아래 마크다운 셀에 분석 내용 작성 (2~3줄)

#### TODO 4 답안: top_k 차이점 분석

**관찰 결과:**
- top_k=2: (질문과 연관성이 매우 높은 핵심 문서만 정확하게 검색되지만, 응답 생성에 필요한 유용한 문맥 정보가 일부 누락될 수 있습니다.)
- top_k=5: (검색 범위가 넓어져 더 풍부한 관련 문맥 정보를 수집할 수 있으나, 상대적으로 연관성이 낮은 노이즈 문서가 포함될 가능성이 커집니다.)

**결론**: (top_k가 커질수록 검색의 재현율(Recall)이 높아져 풍부한 정보 활용이 가능하지만, 너무 높은 경우 노이즈 증가로 인해 생성 모델의 정밀도(Precision)가 떨어질 수 있으므로 적절한 K값 선정이 중요합니다.)

In [ ]:
# (서술형 TODO) 위 마크다운 셀에 답안을 작성하세요.
# 이 셀은 실행하지 않아도 됩니다.

### TODO 5(서술형): 최종 설정 선택 및 근거 서술

- **요구사항**: Yes24 고객 서비스 RAG 시스템에 가장 적합한 설정을 선택하고 근거 제시
- **작성**: chunk_size, overlap, top_k 각각에 대해 선택한 값과 이유를 3~5줄로 작성

#### TODO 5 답안: 최종 설정 선택 및 근거

**선택한 설정:**
- **chunk_size**: 500
- **chunk_overlap**: 50
- **top_k**: 3

**선택 근거:**

1. **chunk_size 선택 이유**: (Yes24 CS 문서(FAQ, 안내사항)는 1~2개 문단 내외의 짧고 명확한 정보가 많으므로, 청크 크기를 500으로 설정하여 문맥 파괴 없이 답변에 꼭 필요한 지식 단위만 정밀하게 벡터화했습니다.)

2. **chunk_overlap 선택 이유**: 전체 chunk_size의 약 10% 수준인 50으로 설정하여, 문단을 분할하는 과정에서 문장의 문맥이나 핵심 키워드(상품명, 환불 조건 등)가 끊겨 정보가 손실되는 현상을 방지했습니다.

3. **top_k 선택 이유**: CS 시스템 특성상 모호한 여러 정보보다 정확한 1~2개의 답변이 중요하므로, top_k를 3으로 설정하여 답변 생성에 필요한 문맥을 충분히 확보하면서도 불필요한 노이즈 문서의 유입을 최소화했습니다.

In [15]:
# 최종 설정으로 RAG 파이프라인 구성
# TODO: 위에서 선택한 설정으로 아래 값들을 수정하세요

FINAL_CHUNK_SIZE = 500      # 1. 선택한 값으로 수정
FINAL_CHUNK_OVERLAP = 50    # 2. 선택한 값으로 수정
FINAL_TOP_K = 3             # 3. 선택한 값으로 수정

final_vectorstore, final_chunks = create_vectorstore_with_chunking(
    documents=all_documents,
    chunk_size=FINAL_CHUNK_SIZE,
    chunk_overlap=FINAL_CHUNK_OVERLAP,
    collection_name="yes24_final"
)

final_retriever = create_rag_components(final_vectorstore, k=FINAL_TOP_K)

print("✅ 최종 RAG 파이프라인 구성 완료")
print(f"📊 설정: chunk_size={FINAL_CHUNK_SIZE}, overlap={FINAL_CHUNK_OVERLAP}, top_k={FINAL_TOP_K}")
print(f"📊 총 청크 수: {len(final_chunks)}개")

✅ 최종 RAG 파이프라인 구성 완료
📊 설정: chunk_size=500, overlap=50, top_k=3
📊 총 청크 수: 33개


In [16]:
# 최종 파이프라인 테스트
test_questions = [
    "총알배송이 뭔가요?",
    "배송이 늦으면 보상받을 수 있나요?",
    "신규 회원 혜택은 무엇인가요?"
]

print("🧪 최종 RAG 파이프라인 테스트")
print("="*60)

for q in test_questions:
    print(f"\n📝 질문: {q}")
    print("\n🤖 응답:")
    response = run_rag(final_retriever, q)
    print(response)
    print("-"*60)

🧪 최종 RAG 파이프라인 테스트

📝 질문: 총알배송이 뭔가요?

🤖 응답:
총알배송은 Yes24의 '빠른 배송' 서비스입니다. 검색된 자료 기준으로 안내해 드리면:

- 당일배송 표시 상품을 주문하시면 같은 날 받아보실 수 있습니다. 서비스 지역은 점차 확대되고 있습니다.  
- 지역(예: 서울·수도권·인천·대전·세종 등)에 따라 당일배송 가능 여부와 예상 도착 시간이 다릅니다.  
- 사무실 등 받는 분이 퇴근하는 장소로 주문하면 당일배송이 어려울 수 있으니, 오후 늦게라도 수령 가능한 주소를 입력하시면 원활합니다.  
- 참고로 '하루배송' 표시 상품은 주문하면 다음 날 받아보실 수 있습니다.  
- 일부 지역은 공휴일(일요일 제외)에도 당일배송이 제공됩니다.

더 궁금하신 점이나 배송 가능 여부(주소별 확인)를 원하시면 배송지 정보를 알려주시면 확인해 드리겠습니다.
------------------------------------------------------------

📝 질문: 배송이 늦으면 보상받을 수 있나요?

🤖 응답:
네, 조건을 충족하면 보상받으실 수 있습니다. 주요 내용은 다음과 같습니다.

- 보상 대상: 예스24 직배송의 아침/당일/하루 배송 대상 상품
- 보상 기준: 주문 시 안내된 도착예상일보다 지연된 경우
  - (원문 기준 예시) 아침 = 도착예상일 아침 7시 이후 지연, 당일/하루 = 도착예상일+1일 이상 지연
  - 단, 주문취소·반품 처리된 주문은 제외
- 보상금액: 주문 건당 YES포인트 2,000원
- 신청 기간: 상품의 주문 상태가 출고 완료 상태일 때, 출고 완료일로부터 14일 이내에 신청 가능 (15일 이상 경과 시 신청 불가)
- 지급 시기: 신청 후 익일에 YES포인트 자동 지급(다만 택배사 확인 지연으로 지급이 지연될 수 있음)
- 제외 사항: 비회원, 아침/당일/하루 배송 대상이 아닌 주문, 고객 사유(부재 등)로 인한 지연, 명절 성수기(설/추석 등) 등 보상 서비스 불가 공지한 경우, 환불·교환 주

---
## 트러블슈팅 가이드

실습/과제 진행 중 자주 발생하는 오류와 해결 방법입니다.

### API 관련

| 증상 | 원인 | 해결 방법 |
|------|------|----------|
| `AuthenticationError` | API Key 미설정 또는 잘못된 키 | `.env` 파일에 `GMS_KEY` 확인 |
| `RateLimitError` | API 호출 한도 초과 | 잠시 대기 후 재실행 (1~2분) |
| `InvalidRequestError` | 토큰 제한 초과 | 입력 텍스트 길이 줄이기 또는 청킹 크기 조정 |

### 패키지 관련

| 증상 | 원인 | 해결 방법 |
|------|------|----------|
| `ModuleNotFoundError` | 패키지 미설치 | Step 1의 `%pip install` 셀 재실행 |
| `ImportError: cannot import name` | 버전 불일치 | `%pip install --upgrade <패키지>` |
| ChromaDB `sqlite3` 오류 | SQLite 버전 낮음 | `%pip install pysqlite3-binary` 후 커널 재시작 |

### 일반

| 증상 | 원인 | 해결 방법 |
|------|------|----------|
| `NameError: name 'xxx' is not defined` | 이전 셀 미실행 | Step 1부터 순서대로 재실행 |
| 출력이 비어있음 | 환경 변수 로드 실패 | `.env` 파일 경로 확인 후 커널 재시작 |


---
# 과제 마무리

## 학습 내용 정리

이번 과제에서 배운 내용:

1. **Step 2**: chunk_size 변경에 따른 검색 품질 차이 체감
2. **Step 3**: top_k 변경에 따른 RAG 응답 품질 차이 체감
3. **Step 4**: 상황에 맞는 설정 선택 및 근거 제시

## 학생용 자가 체크리스트

- [ ] chunk_size 변경에 따른 검색 결과 차이를 설명할 수 있다
- [ ] top_k 변경에 따른 RAG 응답 차이를 설명할 수 있다
- [ ] 선택한 설정에 대한 논리적 근거를 제시할 수 있다
- [ ] 모든 TODO를 완료했다
- [ ] 코드가 에러 없이 실행된다

---

### **Content License Agreement**

<font color='red'><b>**WARNING**</b></font> : 본 자료는 삼성 청년 SW아카데미의 컨텐츠 자산으로, 보안서약서에 의거하여 어떠한 사유로도 임의로 복사, 촬영, 녹음, 복제, 보관, 전송하거나 허가 받지 않은 저장매체를 이용한 보관, 제3자에게 누설, 공개 또는 사용하는 등의 무단 사용 및 불법 배포 시 법적 조치를 받을 수 있습니다.